modules/core/src/persistence.cpp

```diff
@@ -809,56 +792,70 @@ class FileStorage::Impl : public FileStorage_API
+    }
+
+    char* gets( size_t maxCount )
     {
         if( strbuf )
         {
             size_t i = strbufpos, len = strbufsize;
-            int j = 0;
             const char* instr = strbuf;
-            while( i < len && j < maxCount-1 )
+            for( ; i < len; i++ )
             {
-                char c = instr[i++];
-                if( c == '\0' )
-                    break;
-                str[j++] = c;
-                if( c == '\n' )
+                char c = instr[i];
+                if( c == '\0' || c == '\n' )
+                {
+                    if( c == '\n' )
+                        i++;
                     break;
+                }
             }
-            str[j++] = '\0';
+            size_t count = i - strbufpos;
+            if( maxCount == 0 || maxCount > count )
+                maxCount = count;
+            buffer.resize(std::max(buffer.size(), maxCount + 8));
+            memcpy(&buffer[0], instr + strbufpos, maxCount);
+            buffer[maxCount] = '\0';
             strbufpos = i;
-            if (maxCount > 256 && !(flags & cv::FileStorage::BASE64))
-                CV_Assert(j < maxCount - 1 && "OpenCV persistence doesn't support very long lines");
-            return j > 1 ? str : 0;
+            return maxCount > 0 ? &buffer[0] : 0;
         }
-        if( file )
-        {
-            char* ptr = fgets( str, maxCount, file );
-            if (ptr && maxCount > 256 && !(flags & cv::FileStorage::BASE64))
-            {
-                size_t sz = strnlen(ptr, maxCount);
-                CV_Assert(sz < (size_t)(maxCount - 1) && "OpenCV persistence doesn't support very long lines");
-            }
-            return ptr;
-        }
-#if USE_ZLIB
-        if( gzfile )
+
+        const size_t MAX_BLOCK_SIZE = INT_MAX/2; // hopefully, that will be enough
+        if( maxCount == 0 )
+            maxCount = MAX_BLOCK_SIZE;
+        else
+            CV_Assert(maxCount < MAX_BLOCK_SIZE);
+        size_t ofs = 0;
+
+        for(;;)
         {
-            char* ptr = gzgets( gzfile, str, maxCount );
-            if (ptr && maxCount > 256 && !(flags & cv::FileStorage::BASE64))
-            {
-                size_t sz = strnlen(ptr, maxCount);
-                CV_Assert(sz < (size_t)(maxCount - 1) && "OpenCV persistence doesn't support very long lines");
-            }
-            return ptr;
+            int count = (int)std::min(buffer.size() - ofs - 16, maxCount);
+            char* ptr = getsFromFile( &buffer[ofs], count+1 );
+            if( !ptr )
+                break;
+            int delta = (int)strlen(ptr);
+            ofs += delta;
+            maxCount -= delta;
+            if( ptr[delta-1] == '\n' || maxCount == 0 )
+                break;
+            if( delta == count )
+                buffer.resize((size_t)(buffer.size()*1.5));
         }
-#endif
-        CV_Error(CV_StsError, "The storage is not opened");
+        return ofs > 0 ? &buffer[0] : 0;
```

Comment:
yes. I'm sure. when ".gets()" is called by the XML/YAML/JSON parser, it returns the pointer to the read line. The parsers process the line and then call ".gets()" another time to get the next line and so on. Of course, they do not store pointers to the middle of the previously read line, because they consume all the content. ".resize()" can be safely called inside ".gets()".